Passphrase-Protected Key Wrapping

soatok opened the pull request:
The latest OPAQUE draft is significantly improved from earlier designs.

Open question to be decided:

We currently propose a simple construction ~~(AES-256-CTR + HMAC-SHA384, encrypt-then-MAC)~~ to actually encrypt the keys with the OPAQUE export_key.

~~Would ChaCha20 + BLAKE2b-MAC be better? (AES has cache-timing attacks.)~~

~~Are we going to double down on libsodium for our underlying library?~~

~~(To decide in a future commit.)~~


EDIT: Now using ChaCha20 + BLAKE2b-MAC instead of AES-CTR + HMAC. https://github.com/soatok/mastodon-e2ee-specification/pull/4/commits/ea1f59308aaebe999a705ac71377e7a9cbf0bbc1

soatok committed:
Passphrase-Protected Key Wrapping

The latest OPAQUE draft is significantly improved from earlier designs.

Open question to be decided:

We currently propose a simple construction (AES-256-CTR + HMAC-SHA384, encrypt-then-MAC) to actually encrypt the keys
with the OPAQUE export_key.

Would ChaCha20 + BLAKE2b-MAC be better? (AES has cache-timing attacks.)

Are we going to double down on libsodium for our underlying library?

(To decide in a future commit.)

BlackthornYugen commented on components/secret-keys/passphrase-protect.md:
```diff
@@ -1 +1,250 @@
+highly recommended for most users that elect this key recovery mechanism.
+
+For simplicity, we are going to implement Draft 09 or later of the 
+[OPAQUE RFC](7https://datatracker.ietf.org/doc/draft-irtf-cfrg-opaque/09/). OPAQUE will be used to store and recover
```
Typo? 

```diff
-[OPAQUE RFC](7https://datatracker.ietf.org/doc/draft-irtf-cfrg-opaque/09/)
+[OPAQUE RFC](https://datatracker.ietf.org/doc/draft-irtf-cfrg-opaque/09/)
```

soatok replied:
Thanks!

soatok committed:
Fix typo

BlackthornYugen commented on components/secret-keys/passphrase-protect.md:
```diff
@@ -1 +1,250 @@
+
+**Outputs**:
+
+* `export_key` - OPAQUE export key
```
```suggestion
* `main_key` - plaintext Main Key
```

soatok committed:
Update components/secret-keys/passphrase-protect.md

Co-authored-by: John Steel <git@jskw.dev>

BlackthornYugen reviewed (approved):
Looks good 😸

soatok committed:
Use ChaCha20/BLAKE2b instead of AES/SHA2

Since this is my proposal, I'm going to make an executive decision right now to eschew AES from the design

- ChaCha is safer to implement in software than AES
- The 256-bit birthday bound for ChaCha avoids many issues that plague AES-based protocols
- We aren't ever planning to sell anything, let alone to the US government, so FIPS-compatible algorithms don't really matter to us at all
- Libsodium provides a safe and secure implementation of the same cryptographic primitives for many languages, which will aid implementors

We're still using HKDF-SHA512 because that's specified elsewhere. I might drop it for libsodium's crypto_kdf_derive_from_key() in a future commit.

soatok commented:
Merging as-is. This ended up being less controversial than I anticipated before I dived deeper into the latest OPAQUE drafts.
